### DISCUSSION TITLE
Semantic HTML and Web components

### DISCUSSION THREAD

**Discussion by steve2507** [lit/lit#3534] [General]
This may not be a lit question per se, but I'm asking it here anyway since lit is 'my' door to the web component world.

Whenever specifying custom components which have a semantic meaning fitting existing standardized HTML elements, do we still rely on those existing elements?

## non web-components
As an example; I would like to render a list of cards. Traditionally, the implement could look like this;
- ul.cards
  - li

## web components
In the web-component world, I see several possibilities.
Let's say we have a `whatever-card-list` element, which fetches data using [Task](https://github.com/lit/lit/tree/main/packages/labs/task) to render a list;

### nesting the custom component in a semantic list
- whatever-card-list
  - ul
    - li
      - whatever-card-item

### nest the custom component as part of a defined list.
In this example, the `whatever-card-item` replaces the `li` element.
Not sure if this is OK per HTML spec. Probably not.
- whatever-card-list
  - ul
    - whatever-card-item

### do not rely on pre-existing semantic components at all.
- whatever-card-list
  - whatever-card-item

I am wondering if there are any best practices here.

**Comment by bicknellr** [comment]
It's usually best to use the built-in elements if you can so that you don't need to reimplement their accessibility behavior for your own component. I'd recommend putting the `<ul>` inside the `<whatever-card-list>` and `<li>` in the `<whatever-card-item>`s, [like this](https://lit.dev/playground/#project=W3sibmFtZSI6IndoYXRldmVyLWNhcmQtbGlzdC5qcyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcblxuZXhwb3J0IGNsYXNzIFdoYXRldmVyQ2FyZExpc3QgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IFtjc3NgXG4gICAgOmhvc3Qge1xuICAgICAgZGlzcGxheTogYmxvY2s7XG4gICAgICBib3JkZXI6IDJweCBkYXNoZWQgYmx1ZTtcbiAgICB9XG4gIGBdO1xuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGA8dWw-PHNsb3Q-PC9zbG90PjwvdWw-YDtcbiAgfVxufVxuY3VzdG9tRWxlbWVudHMuZGVmaW5lKCd3aGF0ZXZlci1jYXJkLWxpc3QnLCBXaGF0ZXZlckNhcmRMaXN0KTtcbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi93aGF0ZXZlci1jYXJkLWxpc3QuanNcIj48L3NjcmlwdD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi93aGF0ZXZlci1jYXJkLWl0ZW0uanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8d2hhdGV2ZXItY2FyZC1saXN0PlxuICAgIDx3aGF0ZXZlci1jYXJkLWl0ZW0-Rmlyc3QgSXRlbTwvd2hhdGV2ZXItY2FyZC1pdGVtPlxuICAgIDx3aGF0ZXZlci1jYXJkLWl0ZW0-U2Vjb25kIEl0ZW08L3doYXRldmVyLWNhcmQtaXRlbT5cbiAgICA8d2hhdGV2ZXItY2FyZC1pdGVtPlRoaXJkIEl0ZW08L3doYXRldmVyLWNhcmQtaXRlbT5cbiAgPC93aGF0ZXZlci1jYXJkLWxpc3Q-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX0seyJuYW1lIjoid2hhdGV2ZXItY2FyZC1pdGVtLmpzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgY3NzLCBMaXRFbGVtZW50fSBmcm9tICdsaXQnO1xuXG5leHBvcnQgY2xhc3MgV2hhdGV2ZXJDYXJkSXRlbSBleHRlbmRzIExpdEVsZW1lbnQge1xuICBzdGF0aWMgc3R5bGVzID0gW2Nzc2BcbiAgICA6aG9zdCB7XG4gICAgICBkaXNwbGF5OiBjb250ZW50cztcbiAgICB9XG5cbiAgICA6aG9zdCg6aG92ZXIpIHtcbiAgICAgIGNvbG9yOiByZWQ7XG4gICAgfVxuICBgXTtcblxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgPGxpPjxzbG90Pjwvc2xvdD48L2xpPmA7XG4gIH1cbn1cbmN1c3RvbUVsZW1lbnRzLmRlZmluZSgnd2hhdGV2ZXItY2FyZC1pdGVtJywgV2hhdGV2ZXJDYXJkSXRlbSk7XG4ifV0).

If you find yourself in a situation where you can't use the built-in elements for whatever reason, definitely look through [the W3C's ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/). It describes how [common UI patterns](https://www.w3.org/WAI/ARIA/apg/patterns/) should be assembled in the DOM (with links to examples) and how they're expected to react to things like keyboard events, etc. Following those guidelines to make sure that your component behaves as expected is very important: otherwise, [it can be difficult or even impossible for users that rely on accessibility tools to interact with your page properly](https://www.w3.org/WAI/ARIA/apg/practices/read-me-first/). The [structural roles page](https://www.w3.org/WAI/ARIA/apg/practices/structural-roles/) also has a table that maps between ARIA roles and the built-in elements that they mirror, which is useful if there's a built-in element you know you want to emulate but don't know which role to use.

**Comment by steve2507** [comment]
Thank you @bicknellr.
If I am not mistaken, your solution would render as such;
- whatever-card-list
  - ul
    - whatever-card-item
      - li 

The problem is that the web standards do not allow for nesting arbitrary elements in ul/ol;
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul

> Permitted content: Zero or more `li`, `script` and `template` elements.

I have a feeling however that some of those contents (at least on MDN) are not yet updated for the shadow DOM world.

**Comment by lkraav** [comment]
I would roll with a framework that already ships accessibility for their base components. For customs, follow their model and helpers.

Such as https://github.com/vaadin/web-components
